core/src/main/scala/tapir/generic/OneOfMacro.scala

```diff
@@ -0,0 +1,68 @@
+package tapir.generic
+
+import tapir.SchemaFor
+
+import scala.annotation.tailrec
+import scala.reflect.macros.blackbox
+
+object OneOfMacro {
+  // http://onoffswitch.net/extracting-scala-method-names-objects-macros/
+
+  def oneOfMacro[E: c.WeakTypeTag, V: c.WeakTypeTag](
+      c: blackbox.Context
+  )(extractor: c.Expr[E => V], asString: c.Expr[V => String])(mapping: c.Expr[(V, SchemaFor[_])]*): c.Expr[SchemaFor[E]] = {
+    import c.universe._
+
+    @tailrec
+    def resolveFunctionName(f: Function): String = {
+      f.body match {
+        // the function name
+        case t: Select => t.name.decodedName.toString
+
+        case t: Function =>
+          resolveFunctionName(t)
+
+        // an application of a function and extracting the name
+        case t: Apply if t.fun.isInstanceOf[Select] =>
+          t.fun.asInstanceOf[Select].name.decodedName.toString
+
+        // curried lambda
+        case t: Block if t.expr.isInstanceOf[Function] =>
+          val func = t.expr.asInstanceOf[Function]
+
+          resolveFunctionName(func)
+
+        case _ =>
+          throw new RuntimeException("Unable to resolve function name for expression: " + f.body)
+      }
+    }
+
+    val coproducts: Set[Symbol] = {
+      val symbol = weakTypeOf[E].typeSymbol
+      if (!symbol.isClass)
+        c.abort(c.enclosingPosition, "Can only enumerate values of a sealed trait or class.")
+      else if (!symbol.asClass.isSealed)
+        c.abort(c.enclosingPosition, "Can only enumerate values of a sealed trait or class.")
+      else {
+        val children = symbol.asClass.knownDirectSubclasses
+        if (!children.forall(_.isClass))
+          c.abort(c.enclosingPosition, "All children must be objects.")
+        else {
+          children
+        }
+      }
+    }
+
+    val coproductSchemas = coproducts.map { coproduct =>
```

Comment:
why not take `$mapping.value`? :)

Reply:
:rofl:

Reply:
Since we already have this code to find all related classes maybe check the value of `mapping` and fail if there are any missing cases?

Reply:
hmm, well thinking more about it I would definitely use the provided schemas instead of ones found by implicit search (they might be different)

not sure if we'll be able to validate anything except the number of subclasses - how would you know which entry in the schema map corresponds to a specific subclass?

Reply:
> hmm, well thinking more about it I would definitely use the provided schemas instead of ones found by implicit search (they might be different)

+1
> not sure if we'll be able to validate anything except the number of subclasses - how would you know which entry in the schema map corresponds to a specific subclass?

I was thinking about checking the number of subclasses, so we can fail compilation if we detect a non exhaustive match

Reply:
ok, let's try and see if there are bug reports ;-)